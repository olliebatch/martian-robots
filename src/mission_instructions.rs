use crate::robots::{Robot, RobotPosition};
use anyhow::anyhow;
use std::fmt;
use std::str::FromStr;

#[derive(Debug, Clone, PartialEq)]
pub struct Coordinates {
    pub x: i32,
    pub y: i32,
}

impl FromStr for Coordinates {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // remove any spaces if necessary
        let coords: String = s.chars().filter(|c| !c.is_whitespace()).collect();

        if coords.len() != 2 {
            //handle nicer error
            return Err(anyhow!("Too many characters for Coordinates"));
        }
        let mut chars = coords.chars();
        let x_from_char = chars.next().unwrap().to_string().parse::<i32>()?;
        let y_from_char = chars.next().unwrap().to_string().parse::<i32>()?;

        Ok(Coordinates {
            x: x_from_char,
            y: y_from_char,
        })
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Command {
    pub upper_right: Coordinates,
    pub robots: Vec<Robot>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Orientation {
    North,
    South,
    West,
    East,
}

impl fmt::Display for Orientation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let val = match self {
            Orientation::North => "N",
            Orientation::South => "S",
            Orientation::West => "W",
            Orientation::East => "E",
        };
        write!(f, "{:}", val)
    }
}

impl FromStr for Orientation {
    type Err = anyhow::Error;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        match input {
            "N" => Ok(Orientation::North),
            "S" => Ok(Orientation::South),
            "W" => Ok(Orientation::West),
            "E" => Ok(Orientation::East),
            _ => Err(anyhow!("Error matching orientation")),
        }
    }
}
impl Orientation {
    pub fn change_right(&self) -> Self {
        match self {
            Orientation::North => Orientation::East,
            Orientation::East => Orientation::South,
            Orientation::South => Orientation::West,
            Orientation::West => Orientation::North,
        }
    }
    pub fn change_left(&self) -> Self {
        match self {
            Orientation::North => Orientation::West,
            Orientation::East => Orientation::North,
            Orientation::South => Orientation::East,
            Orientation::West => Orientation::South,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum RobotCommands {
    Left,
    Right,
    Forward,
}

impl RobotCommands {
    pub fn process(&self, robot_position: RobotPosition) -> RobotPosition {
        match self {
            RobotCommands::Right => {
                let new_orientation = robot_position.orientation.change_right();
                robot_position.update_orientation(new_orientation)
            }
            RobotCommands::Left => {
                let new_orientation = robot_position.orientation.change_left();
                robot_position.update_orientation(new_orientation)
            }
            RobotCommands::Forward => robot_position.move_forward(),
        }
    }
}

impl FromStr for RobotCommands {
    type Err = anyhow::Error;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        match input {
            "L" => Ok(RobotCommands::Left),
            "R" => Ok(RobotCommands::Right),
            "F" => Ok(RobotCommands::Forward),
            _ => Err(anyhow!("Error matching possible Robot Commands")),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::mission_instructions::{Coordinates, Orientation, RobotCommands};
    use std::str::FromStr;

    use rstest::*;

    #[rstest]
    #[case("1 2", 1, 2)]
    #[case("12", 1, 2)]
    #[case("2 3", 2, 3)]
    #[case("23", 2, 3)]
    fn test_from_str_for_coords(
        #[case] input: &str,
        #[case] expected_x: i32,
        #[case] expected_y: i32,
    ) {
        let p = Coordinates::from_str(input);
        assert_eq!(
            p.unwrap(),
            Coordinates {
                x: expected_x,
                y: expected_y
            }
        )
    }

    #[rstest]
    #[case("N", Orientation::North)]
    #[case("W", Orientation::West)]
    #[case("E", Orientation::East)]
    #[case("S", Orientation::South)]
    fn test_orientation_from_str(#[case] input: &str, #[case] expected_orientation: Orientation) {
        let p = Orientation::from_str(input);
        assert_eq!(p.unwrap(), expected_orientation)
    }

    #[test]
    fn test_orientation_err_from_str() {
        let p = Orientation::from_str("Y");
        assert!(p.is_err())
    }

    #[rstest]
    #[case("F", RobotCommands::Forward)]
    #[case("L", RobotCommands::Left)]
    #[case("R", RobotCommands::Right)]
    fn test_robot_commands(#[case] input: &str, #[case] expected_command: RobotCommands) {
        let command = RobotCommands::from_str(input);

        assert_eq!(command.unwrap(), expected_command);
    }

    #[rstest]
    #[case(Orientation::North, Orientation::East)]
    #[case(Orientation::South, Orientation::West)]
    #[case(Orientation::East, Orientation::South)]
    #[case(Orientation::West, Orientation::North)]
    fn test_moving_right(#[case] input: Orientation, #[case] expected_orientation: Orientation) {
        let p = input.change_right();
        assert_eq!(p, expected_orientation)
    }
    #[rstest]
    #[case(Orientation::North, Orientation::West)]
    #[case(Orientation::South, Orientation::East)]
    #[case(Orientation::East, Orientation::North)]
    #[case(Orientation::West, Orientation::South)]
    fn test_moving_left(#[case] input: Orientation, #[case] expected_orientation: Orientation) {
        let p = input.change_left();
        assert_eq!(p, expected_orientation)
    }
}
