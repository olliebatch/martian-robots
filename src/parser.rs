use crate::mission_instructions::{
    Command, Coordinates, Orientation, RobotCommands, RobotInfo, RobotPosition,
};
use anyhow::anyhow;
use itertools::Itertools;
use std::error;
use std::str::{FromStr, Lines};

pub fn parse_input_to_command(commands: &str) -> Result<Command, Box<dyn error::Error>> {
    let mut lines_of_instruction = commands.lines();
    //todo add a check here to actually check the number of lines etc
    let coords = lines_of_instruction.next();
    // match until cover the lines so that can unwrap.
    //todo add check here to ensure max value is not over 50
    let coordinates = match coords {
        None => Err(anyhow!("No Upper right point provided.")),
        Some(values) => Ok(Coordinates::from_str(values)?),
    }?;

    let robots = parse_robot_commands(lines_of_instruction)?;

    let command = Command {
        upper_right: coordinates,
        robot_commands: robots,
    };

    Ok(command)
}

fn parse_robot_commands(lines: Lines) -> Result<Vec<RobotInfo>, anyhow::Error> {
    let trimmed_lines = remove_lines_and_whitespace(lines);

    // assume current structure will stay the same with 2 lines = one robot
    let robot_infos = generate_robots_from_strs(trimmed_lines)?;

    Ok(robot_infos)
}

fn remove_lines_and_whitespace(lines: Lines) -> Vec<String> {
    //todo if time how to handle a robot that's dropped with no movements?
    let mut removed_lines: Vec<String> = vec![];
    for line in lines {
        let removed_whitespace: String = line.chars().filter(|c| !c.is_whitespace()).collect();

        if removed_whitespace.is_empty() {
            // do nothing because it's an empty line.
        } else {
            removed_lines.push(removed_whitespace)
        }
    }
    removed_lines
}

fn generate_robots_from_strs(
    trimmed_strings: Vec<String>,
) -> Result<Vec<RobotInfo>, anyhow::Error> {
    let mut robots = vec![];
    for chunk in &trimmed_strings.into_iter().chunks(2) {
        let mut robot = RobotInfo::new();
        for (index, robot_info) in chunk.into_iter().enumerate() {
            if index == 0 {
                let coordinates = Coordinates::from_str(&robot_info[..2])?;
                let orientation = Orientation::from_str(&robot_info[2..3])?;
                let robot_position = RobotPosition {
                    coordinates,
                    orientation,
                };
                robot = robot.set_start_position(robot_position);
            } else {
                let robot_commands: Result<Vec<RobotCommands>, anyhow::Error> = robot_info
                    .chars()
                    .map(|test| RobotCommands::from_str(test.to_string().as_str()))
                    .collect();
                robot = robot.update_commands(robot_commands?);
            }
        }
        robots.push(robot)
    }

    Ok(robots)
}

#[cfg(test)]
mod test {
    use crate::parser::{
        generate_robots_from_strs, parse_input_to_command, parse_robot_commands,
        remove_lines_and_whitespace,
    };

    #[test]
    fn test_parse_input_commands() {
        let str = "5 3
        3 2 N
        FRRFLLFFRRFLL";

        let robot_commands = parse_input_to_command(str).unwrap();

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_parse_multiple_commands() {
        let str = "5 3
        1 1 E
        RFRFRFRF

        3 2 N
        FRRFLLFFRRFLL
        
        0 3 W
        LLFFFLFLFL";

        let robot_commands = parse_input_to_command(str).unwrap();

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_parsing_robot_commands() {
        let str = "
        3 2 N
        FRRFLLFFRRFLL";
        let lines = str.lines();

        let robot_commands = parse_robot_commands(lines).unwrap();

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_parsing_multiple_robot_commands() {
        let str = "
        1 1 E
        RFRFRFRF
        3 2 N
        FRRFLLFFRRFLL";
        let lines = str.lines();

        let robot_commands = parse_robot_commands(lines).unwrap();

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_remove_lines_and_whitespace_space_and_lines() {
        let str = "
        3 2 N

        FRRFLLFFRRFLL";
        let lines = str.lines();

        let removed_lines = remove_lines_and_whitespace(lines);
        let expected = vec!["32N".to_string(), "FRRFLLFFRRFLL".to_string()];

        assert_eq!(removed_lines, expected)
    }

    #[test]
    fn test_remove_lines_and_whitespace_no_lines() {
        let str = "
        3 2 N
        FRRFLLFFRRFLL";
        let lines = str.lines();

        let removed_lines = remove_lines_and_whitespace(lines);
        let expected = vec!["32N".to_string(), "FRRFLLFFRRFLL".to_string()];

        assert_eq!(removed_lines, expected)
    }

    #[test]
    fn test_generate_robots_from_strs() {
        let robots =
            generate_robots_from_strs(vec!["32N".to_string(), "FRRFLLFFRRFLL".to_string()])
                .unwrap();

        insta::assert_debug_snapshot!(robots)
    }

    #[test]
    fn test_generate_robots_from_strs_two_robots() {
        let robots = generate_robots_from_strs(vec![
            "11E".to_string(),
            "RFRFRFRF".to_string(),
            "32N".to_string(),
            "FRRFLLFFRRFLL".to_string(),
        ])
        .unwrap();

        insta::assert_debug_snapshot!(robots)
    }
}
