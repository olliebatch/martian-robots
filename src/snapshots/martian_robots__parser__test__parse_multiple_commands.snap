---
source: src/parser.rs
assertion_line: 117
expression: robot_commands
---
Command {
    upper_right: Coordinates {
        x: 5,
        y: 3,
    },
    robot_commands: [
        RobotInfo {
            start_position: RobotPosition {
                coordinates: Coordinates {
                    x: 1,
                    y: 1,
                },
                orientation: East,
            },
            robot_commands: [
                Right,
                Forward,
                Right,
                Forward,
                Right,
                Forward,
                Right,
                Forward,
            ],
        },
        RobotInfo {
            start_position: RobotPosition {
                coordinates: Coordinates {
                    x: 3,
                    y: 2,
                },
                orientation: North,
            },
            robot_commands: [
                Forward,
                Right,
                Right,
                Forward,
                Left,
                Left,
                Forward,
                Forward,
                Right,
                Right,
                Forward,
                Left,
                Left,
            ],
        },
        RobotInfo {
            start_position: RobotPosition {
                coordinates: Coordinates {
                    x: 0,
                    y: 3,
                },
                orientation: West,
            },
            robot_commands: [
                Left,
                Left,
                Forward,
                Forward,
                Forward,
                Left,
                Forward,
                Left,
                Forward,
                Left,
            ],
        },
    ],
}
